### DISCUSSION TITLE
Analysing Model Predictions Post-Training (First Order Derivatives)

### DISCUSSION THREAD

**Discussion by elliot-hicks** [lululxvi/deepxde#914] [Q&A] [answered]
Hi Lu Lu/ All,

Firstly, big fan of DeepXDE, thank you for this great open-source tool! 

Secondly, I was hoping I could get some help from you/ the DeepXDE community. I have been working on a fluid dynamics problem using your library, and have been successful in simulating the stream function $\psi$ for lid-driven flow in a cavity. In this implementation the model inputs are the 2D cartesian coordinates [x, y], and the outputs are $\psi$ and pressure, p:

$[\psi, p]$ = Model(x, y).

The network predictions for $\psi$ and p are great, but I was also hoping to analyse the derivatives of $\psi$ after training the model. (In this case, the gradients of $\psi$ are the velocities $u$ and $v$.) In this implementation, I was hoping to give the model a set of coordinates to predict on after training, and receive [ $\psi$ , p, $\frac{d\psi}{dx} $, $\frac{d\psi}{dy} $]. I have attempted to use the FirstDerivative callback, and output processing also, but have had no successes.

For context, I am using the Tensorflow 1.x backend, and my PDE system is defined as follows:

````
def ns_pde_system(x, y) -> list:
  """

  The 2D Navier Stokes PDE system:

  Inputs:
    x (tuple): (x,y) collocation points in our domain
    y (tuple): (psi, pressure) corresponding to our collocation points

    Outputs:
    [
    continuity: we know as the fluid is incompressible, this is zero: psi_x + psi_y := 0
    x_momentum: dim one of NS - also known to be zero by the PDE
    y_momentum: dim two of NS - also known to be zero by the PDE
    ]
  """

  #2D Navier Stokes:
  # First Order Derivs:
  psi, p = y[:, 0:1], y[:, 1:2]

  u = dde.grad.jacobian(y, x, i=0, j=1)    # u = psi_y
  v = -1*dde.grad.jacobian(y, x, i=0, j=0) # v = -psi_x

  u_x = dde.grad.jacobian(u, x, i=0, j=0)
  u_y = dde.grad.jacobian(u, x, i=0, j=1)

  v_x = dde.grad.jacobian(v, x, i=0, j=0)
  v_y = dde.grad.jacobian(v, x, i=0, j=1)

  u_xx = dde.grad.jacobian(u_x, x, i=0, j=0)
  u_yy = dde.grad.jacobian(u_y, x, i=0, j=1)
  v_xx = dde.grad.jacobian(v_x, x, i=0, j=0)
  v_yy = dde.grad.jacobian(v_y, x, i=0, j=1)
  
  p_x = dde.grad.jacobian(y, x, i=1, j=0)
  p_y = dde.grad.jacobian(y, x, i=1, j=1)
  
  continuity = u_x + v_y
  x_momentum = RHO * (u * u_x + v * u_y) + p_x - NU * (u_xx + u_yy)
  y_momentum = RHO * (u * v_x + v * v_y) + p_y - NU * (v_xx + v_yy)

  return [continuity, x_momentum, y_momentum]`
```


I understand many implementations work directly with u and v, but in this experiment I was hoping to work with 
the stream function $\psi$.

Thanks again for the great library, and in advance of any help this forum may provide.

**Comment by lululxvi** [comment, marked as answer]
FirstDerivative callback is the correct way to get first order derivatives of \psi.

**Reply by elliot-hicks** [reply]
Okay thank you, I'll try to get it to work.
